### PR TITLE
Refactoring code

### PR DESCRIPTION
**Pull request by idali0226**

### CODE REVIEW

File: src/components/Filter.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import PropTypes from 'prop-types'
+import Dropdown from 'react-dropdown'
+
+const propTypes = {
+  handleFilterUpdate: PropTypes.func.isRequired,
+  currentFilter: PropTypes.string.isRequired,
+}
+
+class Filter extends React.Component {
+  static getDerivedStateFromProps(nextProps, prevState) {
```
Comment (antonoberg) [inline src/components/Filter.js]: Cool new function

File: src/components/Filter.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  constructor(props) {
+    super(props)
+    this.state = {
+      currentFilter: props.currentFilter,
+    }
+
+    this.handleFilterUpdate = this.handleFilterUpdate.bind(this)
+  }
+
+  handleFilterUpdate(value) {
```
Comment (antonoberg) [inline src/components/Filter.js]: This is not necessary. You can just pass the function directly to the dropdown with  onChange={this.props.handleFilterUpdate}

File: src/components/Filter.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import Dropdown from 'react-dropdown'
+
+const propTypes = {
+  handleFilterUpdate: PropTypes.func.isRequired,
```
Comment (antonoberg) [inline src/components/Filter.js]: To be picky: here its better if the method is called onFilterUpdate

File: src/components/TodoBox.js [hunk 4]
```diff
@@ -127,8 +122,9 @@ export default class TodoBox extends React.Component {
     })
 
     this.setState({
-      selectDefaultValue: 0,
+      currentFilter: 'All',
     })
+    console.log(this.state.currentFilter)
```
Comment (antonoberg) [inline src/components/TodoBox.js]: No console.log. If you run "yarn lint" its complaining on this one

File: src/components/TodoBox.js [hunk 5]
```diff
@@ -151,33 +147,26 @@ export default class TodoBox extends React.Component {
...
             <hr />
-            <h3 className="todo-count">TODO List [{state.todos.length}]</h3>
-            <Dropdown
-              options={options}
-              onChange={this.onSelect}
-              value={defaultOption}
-              placeholder="Select an option"
+            <h3 className="todo-count">TODO List [{todos.length}]</h3>
+            <Filter
+              handleFilterUpdate={this.handleFilterUpdate}
```
Comment (antonoberg) [inline src/components/TodoBox.js]: Here the correct naming would be onFilterUpdate={this.handleFilterUpdate}

File: src/components/EditTodo.js [hunk 6]
```diff
@@ -11,34 +11,17 @@ const propTypes = {
 }
 
 class EditTodo extends React.Component {
-  constructor(props) {
```
Comment (antonoberg) [inline src/components/EditTodo.js]: Nice cleanup.

File: src/components/Filter.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+  render() {
+    const filterOptions = ['All', 'New', 'Done']
+
+    return (
+      <div>
+        <Dropdown
+          options={filterOptions}
+          onChange={this.handleFilterUpdate}
+          value={this.state.currentFilter}
```
Comment (antonoberg) [inline src/components/Filter.js]: I prefere unpacking the state at the top of the render function like:

const {
  currentFilter
} = this.state

File: .eslintrc [hunk 8]
```diff
@@ -26,6 +26,7 @@
     "import/prefer-default-export": 0,
     "indent": 0,
     "jsx-a11y/anchor-is-valid": 0,
+    "react/prefer-stateless-function": [0, { "ignorePureComponents": true }],
```
Comment (antonoberg) [inline .eslintrc]: Very good. This rule makes refactoring harder so good that its disabled

File: src/components/TodoForm.js [hunk 9]
```diff
@@ -45,14 +43,13 @@ class TodosForm extends React.Component {
...
-        isEdit: false,
       })
     }
   }
 
   render() {
-    const { description, name, status, buttonText, isEdit } = this.state
-
+    const { description, name, status, isEdit } = this.state
+    const buttonText = isEdit ? 'Update' : 'Create TODO Item'
```
Comment (antonoberg) [inline src/components/TodoForm.js]: Always good to keep the stuff kept in the state to a minimum

File: src/components/CreateTodo.js [hunk 10]
```diff
@@ -7,35 +7,20 @@ const propTypes = {
...
-      name: '',
-      description: '',
-      status: '',
-    }
-
-    this.addTodo = this.addTodo.bind(this)
-  }
-
-  addTodo(name, description) {
+  createTodo(name, description) {
```
Comment (antonoberg) [inline src/components/CreateTodo.js]: Its good to use default values instead. In this case i would do createTodo({name, description})

In general I think its better to use an object as a function argument as soon as you have more than one argument or if you dont know if you will need to pass more later.

In some cases: For example. setValue(path, value) it makes sense to not use an object but usually an object saves time later on. Its also easier to understand how a function works if you are forced to call it with specific parameterNames as you have to when its an object

File: src/components/TodoForm.js [hunk 11]
```diff
@@ -21,7 +21,7 @@ class TodosForm extends React.Component {
     super(props)
     this.state = {
       isEdit: this.props.isEdit,
-      id: this.props.isEdit ? this.props.id : '',
+      id: this.props.isEdit ? this.props.id : 0,
```
Comment (antonoberg) [inline src/components/TodoForm.js]: Good catch

File: src/components/EditTodo.js [hunk 12]
```diff
@@ -36,7 +37,7 @@ class EditTodo extends React.Component {
           name={state.name}
           description={state.description}
           status={state.status}
-          editTodo={this.updateTodo}
+          onUpdate={this.onUpdate}
```
Comment (antonoberg) [inline src/components/EditTodo.js]: You can pass this method directly from the parent like onUpdate={this.prop.onUpdate}

File: src/components/TodoBox.js [hunk 13]
```diff
@@ -165,34 +151,43 @@ export default class TodoBox extends React.Component {
   }
 
   render() {
-    const todoList = this.getTodos()
+    const { state } = this
```
Comment (antonoberg) [inline src/components/TodoBox.js]: Here you could unpack state further

File: src/components/TodoBox.js [hunk 14]
```diff
@@ -165,34 +151,43 @@ export default class TodoBox extends React.Component {
...
-            </h3>
+            <h3 className="todo-count">TODO List [{state.todos.length}]</h3>
             <Dropdown
               options={options}
               onChange={this.onSelect}
               value={defaultOption}
               placeholder="Select an option"
             />
-            <div>{todoList}</div>
+            <div>
```
Comment (antonoberg) [inline src/components/TodoBox.js]: Detail but i guess this div is not necessary

File: src/components/TodoForm.js [hunk 15]
```diff
@@ -55,15 +51,15 @@ class TodosForm extends React.Component {
   }
 
   render() {
-    const { state } = this
+    const { description, name, status, buttonText } = this.state
```
Comment (antonoberg) [inline src/components/TodoForm.js]: Good unpacking

File: src/components/TodoView.js [hunk 16]
```diff
@@ -0,0 +1,68 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import RemoveConfirmation from './RemoveConfirmation'
+
+const propTypes = {
+  id: PropTypes.number.isRequired,
+  name: PropTypes.string.isRequired,
+  description: PropTypes.string.isRequired,
```
Comment (antonoberg) [inline src/components/TodoView.js]: Is description required?

Only put the required props as required

File: src/components/TodoView.js [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+    this.props.onStatusChange(this.props.id, newStatus)
+  }
+
+  onEdit = e => {
+    e.preventDefault()
+    this.props.onEdit()
+  }
+
+  onDelete() {
```
Comment (antonoberg) [inline src/components/TodoView.js]: This should be handleDelete

File: src/components/TodoView.js [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+          <p className="todo-body">{description}</p>
+          <p className="todo-status">
+            {status}
+            {'   '}
+            <a href="" onClick={this.onStatusChange}>
+              Change status
+            </a>
+          </p>
+          <div className="todo-actions">
+            <RemoveConfirmation onDelete={this.onDelete} />
```
Comment (antonoberg) [inline src/components/TodoView.js]: This should be <RemoveConfirmation onDelete={this.handleDelete} />

File: src/components/TodoItem.js [hunk 19]
```diff
@@ -0,0 +1,83 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import TodoView from './TodoView'
+import EditTodo from './EditTodo'
+
+const propTypes = {
+  id: PropTypes.number.isRequired,
+  name: PropTypes.string.isRequired,
+  description: PropTypes.string.isRequired,
```
Comment (antonoberg) [inline src/components/TodoItem.js]: Not required

File: src/components/TodoItem.js [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+  onStatusChange(id, newStatus) {
+    this.props.onStatusChange(id, newStatus)
+
+    this.setState({
+      isEdit: false,
+    })
+  }
+
+  render() {
```
Comment (antonoberg) [inline src/components/TodoItem.js]: This render method looks very clean now!

File: .eslintrc [hunk 21]
```diff
@@ -13,7 +13,10 @@
     "prettier"
   ],
   "globals": {
-    "fetch": false
+    "document": true,
```
Comment (antonoberg) [inline .eslintrc]: good

File: src/components/Filter.js (hunk 7 above)
Comment (idali0226) [reply src/components/Filter.js]: e13d2e2

File: src/components/TodoView.js (hunk 16 above)
Comment (idali0226) [reply src/components/TodoView.js]: 809ac17

File: src/components/TodoView.js (hunk 17 above)
Comment (idali0226) [reply src/components/TodoView.js]: 3c3176a

File: src/components/TodoView.js (hunk 18 above)
Comment (idali0226) [reply src/components/TodoView.js]: 3c3176a

File: src/components/CreateTodo.js [hunk 22]
```diff
@@ -7,40 +7,20 @@ const propTypes = {
...
-          id={state.id}
-          name={state.name}
-          description={state.description}
-          status={state.status}
+          createTodo={this.props.createTodo}
+          isEdit={false}
+          id={0}
+          name=""
+          description=""
+          status=""
```
Comment (fredrikolovsson) [inline src/components/CreateTodo.js]: id, name, description and status should not be be passed at all to TodoForm. Set defaultProps in TodoForm instead

File: src/components/TodoBox.js [hunk 23]
```diff
@@ -110,7 +72,36 @@ export default class TodoBox extends React.Component {
...
+    } else if (todosCount === 1) {
+      this.title = '1 todo item'
+    } else {
+      this.title = `${todosCount} todo items`
+    }
+    return this.title
+  }
+
+  handleFilterUpdate(value) {
+    this.fetchTodos(value.value)
```
Comment (fredrikolovsson) [inline src/components/TodoBox.js]: looks like you could deconstruct the value from value in the function signature like this: `handleFilterUpdate({ value })`

File: src/components/TodoForm.js [hunk 24]
```diff
@@ -2,72 +2,61 @@ import React from 'react'
...
+  onUpdate: PropTypes.func,
   isEdit: PropTypes.bool.isRequired,
-  id: PropTypes.number,
-  name: PropTypes.string,
-  description: PropTypes.string,
-  status: PropTypes.string,
+  id: PropTypes.number.isRequired,
+  name: PropTypes.string.isRequired,
+  description: PropTypes.string.isRequired,
+  status: PropTypes.string.isRequired,
```
Comment (fredrikolovsson) [inline src/components/TodoForm.js]: those four should not be required, set them to undefined as default

File: src/components/TodoForm.js (hunk 11 above)
Comment (fredrikolovsson) [reply src/components/TodoForm.js]: id should be undefined if it's not in edit mode

**Pull request merged**